### PR TITLE
Organize fetch and checks and import bundles

### PR DESCRIPTION
**Pull request by Gabib32**
The architecture for DIVA is to separate the fetching
of data from the different checks. This has been
accomplished in this commit.

The repo and bundle data, when fetched, is downloaded and
imported into the databse. The update content still needs
this functionality. There are different data structs that
populate the needed information from the database, so the
check commands are not downloading any content.

Signed-off-by: gabrielle beyer <gabrielle.n.beyer@intel.com>

### CODE REVIEW

File: bundle/bundle.go [hunk 1]
```diff
@@ -50,8 +48,69 @@ type Definition struct {
...
 }
 
-// Set is a map of bundle names to their definition
-type Set map[string]*Definition
+// DefinitionsSet is a map of bundle names to their definition
+type DefinitionsSet map[string]*Definition
+
+//
+// // Definitions is a slice of Definition objects
+// type Definitions []*Definition
```
Comment (matthewrsj) [inline bundle/bundle.go]: remove commented-out code please

File: bundle/bundle.go [hunk 2]
```diff
@@ -50,8 +48,69 @@ type Definition struct {
...
+
+	allIncludes := make(map[string]bool)
+	for _, bundle := range defs {
+		for j := range bundle.Includes {
+			allIncludes[j] = true
+		}
+	}
+	var err error
+	if len(allIncludes) == 0 {
+		err = fmt.Errorf("Unable to get all includes, set is empty")
```
Comment (matthewrsj) [inline bundle/bundle.go]: are we assuming that there will always be one include (like `os-core`)?

Comment (matthewrsj) [reply bundle/bundle.go]: But what about os-core? that doesn't have any includes and that isn't an error.

Comment (matthewrsj) [reply bundle/bundle.go]: good test case to add ;)

File: bundle/bundle.go [hunk 3]
```diff
@@ -50,8 +48,69 @@ type Definition struct {
...
+
+	return allIncludes, err
+}
+
+// GetAllPackages returns all of the dependent packages recursively among all
+// bundle definitions. If a bundleName is passed, only the recursive packages
+// to that bundle will be returned.
+func (defs DefinitionsSet) GetAllPackages(bundleName string) (map[string]bool, error) {
+	if bundleName != "" {
+		return defs[bundleName].AllPackages, nil
```
Comment (matthewrsj) [inline bundle/bundle.go]: \o/

File: bundle/bundle.go [hunk 4]
```diff
@@ -50,8 +48,69 @@ type Definition struct {
...
+	allPackages := make(map[string]bool)
+	for _, bundle := range defs {
+		for j := range bundle.AllPackages {
+			allPackages[j] = true
+		}
+	}
+
+	var err error
+	if len(allPackages) == 0 {
+		err = fmt.Errorf("Unable to get all packages, set is empty")
```
Comment (matthewrsj) [inline bundle/bundle.go]: is it invalid to have a bundle definition with no packages (and no includes)?

File: bundle/bundle.go (hunk 2 above)
Comment (Gabib32) [reply bundle/bundle.go]: +1 good catch

Comment (Gabib32) [reply bundle/bundle.go]: ahhhh but os-core will include itself ;)

Comment (Gabib32) [reply bundle/bundle.go]: So boom, it does work :stuck_out_tongue_closed_eyes:

File: bundle/bundle.go (hunk 3 above)
Comment (Gabib32) [reply bundle/bundle.go]: I don't even know why I bother with this, instead of in the code just calling the defs[name].includes ... merp :woman_shrugging:

Comment (Gabib32) [reply bundle/bundle.go]: ^ that's a shrug btw lol

File: bundle/bundle.go (hunk 4 above)
Comment (Gabib32) [reply bundle/bundle.go]: Since they all include themselves, I would think yes. @tmarcu thoughts?

File: bundle/bundle.go (hunk 2 above)
Comment (matthewrsj) [reply bundle/bundle.go]: hmm, I forgot this list is supposed to also list the parent bundle. That's confusing, but ok.

File: bundle/bundle.go (hunk 4 above)
Comment (matthewrsj) [reply bundle/bundle.go]: Not sure I agree with this. I think if a user wants to define an empty OS (no packages even in os-core) they should be allowed to. At least they shouldn't be prohibited by the bundle library. bundle *validation* on the other hand should probably complain.

File: bundle/bundle.go (hunk 3 above)
Comment (matthewrsj) [reply bundle/bundle.go]: ... that's a good point lol

File: bundle/bundle_test.go [hunk 5]
```diff
@@ -588,49 +593,64 @@ func TestGetAllPackagesForOneBundle(t *testing.T) {
...
 	}
 
 	for _, tc := range testCases {
 		// add os-core packages to all bundles
 		for pkg := range core.AllPackages {
-			tc.expectedPackages = append(tc.expectedPackages, pkg)
+			tc.expectedPackages[pkg] = true
+		}
+
+		for _, pkg := range tc.expectedSlice {
```
Comment (matthewrsj) [inline bundle/bundle_test.go]: wait so is the only reason you have expectedSlice so you can add it to the expectedPackages map? Why not just define the expectedPackages as a literal above and add the core.AllPackages like you do?

Comment (Gabib32) [reply bundle/bundle_test.go]: I guess I can, I just already had the slices programmed, so it saved me some of that hardcoding time :) But I'll do that so its consistent throughout

Comment (Gabib32) [reply bundle/bundle_test.go]: lol that took like 2 minutes I was beign so lazy

**Pull request merged**